### PR TITLE
Domain registrant verification landing page

### PR DESCRIPTION
**Pull request by delputnam**
#### Changes proposed in this Pull Request

In order to provide a more consistent and controllable user experience around domains messaging, we are implementing our own domains messaging. Along with that, we also need to provide some simple landing pages for a variety of user interactions such as verifying domain contact information.

In this PR, we want to add the first of these landing pages as well as a framework that will allow us to easily add more domain-messaging related landing pages.

Depends on D26351-code

A successful verification would look like the following:

#### Testing instructions

Apply the patch to the back end and build this branch locally.

Visit: http://calypso.localhost:3000/domain-services/registrant-verification?domain=transfer-cancel-test-2.com&email=5test2@test.com&token=bf772d0c0cf7902917d367d124c74682

This will likely show an error because the contact info for this domain is not in a state where it can be re-verified.

To test the various page states, hack the `/verify-email` endpoint from D26351-code to return each of the following:

`return [ 'success' => true ];` to see the success case:

Screen Shot 2019-04-05 at 1 07 01 PM

`return new WP_Error( 'whois_error', 'There was a problem processing your request, please try again later.', 500 );`
-or-
`return new WP_Error( 'bad_request', 'This is not a proper request.', 400 );` should show a generic error page:

Screen Shot 2019-04-05 at 1 07 35 PM

`return new WP_Error( 'unauthorized', 'You are not allowed to execute this action.', 403 );` indicates that the token has expired and should show an error with an offer to re-send the email with a valid token:

Screen Shot 2019-04-05 at 1 07 54 PM

`return new WP_Error( 'email_mismatch', 'The email you are trying to verify is different from the email we have on record.', 500 );` Should show an error page with some information about the reason for the error:

Screen Shot 2019-04-05 at 1 26 48 PM

Additionally, try visiting the path with an invalid `/:action` (ex.: http://calypso.localhost:3000/domain-services/registrant-verificationzzz?domain=transfer-cancel-test-2.com&email=5test2@test.com&token=bf772d0c0cf7902917d367d124c74682) to make sure you see the equivalent of a 404 page for these domain services landing pages:

Screen Shot 2019-04-05 at 1 08 56 PM

### CODE REVIEW

File: client/document/domains-landing.jsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/**
```
Comment (sgomes) [inline client/document/domains-landing.jsx]: The duplication between this and `index.jsx` worries me. Code tends to have a tendency to grow stale in these situations. In particular, I can see changes to the `evergreen` check being needed as we get more experience with odd customer setups in the wild, and there's a fair chance someone will forget to update one of the files.

I'm not super familiar with this part of the codebase, though, so I'm not entirely sure how best to handle reuse. @blowery: What do you think is the right approach here?

File: server/pages/index.js [hunk 2]
```diff
@@ -171,13 +171,12 @@ const getFilesForChunk = ( chunkName, request ) => {
 	return groupAssetsByType( allTheFiles );
 };
 
-const getFilesForEntrypoint = target => {
-	const entrypointAssets = getAssets( target ).entrypoints.build.assets.filter(
+const getFilesForEntrypoint = ( target, name ) => {
```
Comment (sgomes) [inline server/pages/index.js]: This appears to take into account the changes in the multiple builds PR correctly 👍

File: webpack.config.js [hunk 3]
```diff
@@ -158,7 +158,10 @@ function getWebpackConfig( {
 	const webpackConfig = {
 		bail: ! isDevelopment,
 		context: __dirname,
-		entry: { build: [ path.join( __dirname, 'client', 'boot', 'app' ) ] },
+		entry: {
```
Comment (sgomes) [inline webpack.config.js]: Adding a new entrypoint seems like a pretty fundamental change to how things are currently being done.

@blowery: Does this approach work well with any potential future plans for Calypso around splitting things up into smaller builds?

**Review by sgomes — COMMENTED** [review]
I added some comments related to general code structure. They're mostly questions for @blowery, though! 😁 

From the multiple builds PR point of view, the rebase seems to correctly take the different asset lists into account.

I haven't tested the code, since I'm assuming that's best left for those with deeper knowledge of the domain in question (pun intended). Do let me know if you disagree, and I'll take a look :)
In terms of testing for the multiple builds, it's likely not necessary, but out of an abundance of caution I'd build for production (`CALYPSO_ENV=production`) or use the `calypso.live` URL and perform the same testing on two setups:
- One using a modern browser (e.g. latest Chrome). Ensure JS files come from `/calypso/evergreen/`.
- One using an older browser, or a modern browser with the `forceFallback=1` URL parameter. Ensure JS files come from `/calypso/fallback/`.

File: client/document/domains-landing.jsx (hunk 1 above)
Comment (blowery) [reply client/document/domains-landing.jsx]: I was actively holding off on trying to remove the duplication for a while. I share your concerns about these bits getting stale, but I'd like to get a handle on what the shared bits really are before we start consolidating.

File: webpack.config.js (hunk 3 above)
Comment (blowery) [reply webpack.config.js]: I don't really know, since we don't currently have any solid plans. I don't see it hurting anything though and could open some doors. Each entry point can have its own boot and setup, which may make things easier to isolate.

An enhancement of what we have so far might be to break out a manifest per entry point instead of our current setup with one manifest.

Comment (delputnam) [reply webpack.config.js]: Is this something you think we should address before continuing? Or something we should address in a future PR?

**Comment by delputnam** [comment]
> * One using a modern browser (e.g. latest Chrome). Ensure JS files come from `/calypso/evergreen/`.

Yep. Using newest release of Chrome and Safari, assets come from `/calypso/evergreen/`.

> * One using an older browser, or a modern browser with the `forceFallback=1` URL parameter. Ensure JS files come from `/calypso/fallback/`.

I tested with the `forceFallback` parameter and the assets come from `/calypso/fallback/`.

**Comment by sgomes** [comment]
> Yep. Using newest release of Chrome and Safari, assets come from `/calypso/evergreen/`.

> I tested with the `forceFallback` parameter and the assets come from `/calypso/fallback/`.

All clear from my end, then! 👍 Great work, @delputnam!

**Review by blowery — APPROVED** [review]
Framework parts LGTM.

File: client/landing/domains/registrant-verification/index.jsx [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+				break;
+		}
+
+		this.setState( {
+			...this.getDefaultErrorState(),
+			...errorState,
+		} );
+	};
+
+	resendEmail = () => {
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: One _potential_ problem with this: for the landing page, we don't require the user to match the blog, etc. (or to be logged in at all). However, `wpcom.resendIcannVerification` _does_ require the user to be logged in and be able to manage the domain.

It might be good to check (in a follow-up PR, this one is big enough) to handle the `unknown_blog` error separately (nudging the user to log in, etc.).

Otherwise, it leads to this slightly misleading and confusing screen (when I tried to resend the verification email for a domain I don't own):
calypso localhost_3000_domain-services_registrant-verification_domain=transfer-cancel-test-2 com email=5test2@test com token=bf772d0c0cf7902917d367d124c74682 (1)

File: client/landing/domains/registrant-verification/index.jsx [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		const { domain, translate } = this.props;
+
+		this.setState( this.getLoadingState() );
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: This state does not really apply here - it shows a `Verifying your contact information…` message, but that's not what's happening and the user could be really confused as a result.

Again, not a blocker, should probably be fixed in a separate PR.

File: client/landing/domains/registrant-verification/index.jsx [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+		wpcom.resendIcannVerification( domain, error => {
+			if ( error ) {
+				this.setErrorState( { error: 'resend_email_failed' } );
+			} else {
+				this.setState( {
+					illustration: '/calypso/images/illustrations/illustration-ok.svg',
+					title: translate( 'Email sent!' ),
+					message: translate( 'Check your email.' ),
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: How about we show the email we've sent the email to? We do that in Domain Management and it makes sense here too. File it under "nice to have things" as it needs changes on the backend to return that info (email) from the backend on success.
Not super high-prio, since this whole flow should be an edge case.

File: client/landing/domains/registrant-verification/index.jsx [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+			actionCallback: this.resendEmail,
+		};
+	};
+
+	getEmailMismatchState = () => {
+		const { domain, email, translate } = this.props;
+
+		return {
+			message: translate(
+				'{{strong}}%(email)s{{/strong}} is different from the email we have on record for {{strong}}%(domain)s{{/strong}}.',
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: Maybe let's add a sentence on what to do next? Otherwise, it feels like we're just leaving the user to figure it out themselves.

AFAICT, this state is triggered when the user clicks an old link, so maybe let's nudge the user to check their email for a newer email, etc.

File: client/landing/domains/registrant-verification/index.jsx [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+		};
+	};
+
+	getVerificationSuccessState = () => {
+		const { domain, translate } = this.props;
+		return {
+			illustration: '/calypso/images/illustrations/illustration-ok.svg',
+			title: translate( 'Success!' ),
+			message: translate(
+				'Thank your for verifying your contact information for:{{br /}}{{strong}}%(domain)s{{/strong}}.',
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: The `:` seems a bit clunky (as if we supported verifying more than one domain at once? How about dropping it and making the `strong` `white-space: nowrap` instead?

File: client/landing/domains/content-card/index.jsx [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+			isLoading,
+		} = this.props;
+
+		if ( isLoading ) {
+			return this.renderPlaceholder();
+		}
+
+		return (
+			<CompactCard className="content-card">
+				{ illustration && <img src={ illustration } alt="" /> }
```
Comment (klimeryk) [inline client/landing/domains/content-card/index.jsx]: What do you think about dropping the illustration altogether? It doesn't seem to add any value to the page and it takes up precious space, pushing the most important part, the title, way low. This is especially visible in the loading state (and on mobile devices).

File: client/landing/domains/registrant-verification/index.jsx [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+					},
+					components: {
+						strong: <strong />,
+					},
+				}
+			),
+		};
+	};
+
+	getKeySystemsErrorState = errorMessage => {
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: Feels like we could simply return a proper, unique error key from the backend (either from Domains_API or do the message check in the endpoint). This feels like too much backend logic on the frontend to me 🤔

File: client/landing/domains/registrant-verification/index.jsx [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+			'Invalid attribute value; Contact verification already confirmed, nothing to do' ===
+			errorMessage
+		) {
+			const { domain, email, translate } = this.props;
+
+			return {
+				illustration: '/calypso/images/illustrations/illustration-ok.svg',
+				title: translate( 'Already verified.' ),
+				message: translate(
+					"You've already verified {{strong}}%(email)s{{/strong}} for:{{br /}}{{strong}}%(domain)s{{/strong}}.",
```
Comment (klimeryk) [inline client/landing/domains/registrant-verification/index.jsx]: Same thing for the `:`.

**Review by klimeryk — COMMENTED** [review]
Tested and overall looks great - left some nitpicks. No blockers in particular, but I imagine you'll want to fix some of them anyway 😅

**Review by klimeryk — APPROVED** [review]
Let's merge this as-is (it works 328% better than what our domain partners have anyway) and address the comments in a separate PR to avoid making this PR even bigger and harder to review 👍

File: client/landing/domains/content-card/index.jsx (hunk 9 above)
Comment (aidvu) [reply client/landing/domains/content-card/index.jsx]: And we don't load images. :P

**Review by aidvu — COMMENTED** [review]
Love the work here... Thank you all! 🙇
